### PR TITLE
Android/GL stability, some cocos updates

### PR DESCRIPTION
**Pull request by lagoPS**
- Vertex/Index buffer update bugfixes and safety measures for overflows greatly improve GL version stability
- Some of the new cocos commits merged manually
- Many GL errors fixed mainly due to vertex/index buffers misbehaviours

### CODE REVIEW

File: cocos2d/cocos/2d/CCDrawNode.cpp [hunk 1]
```diff
@@ -112,8 +111,7 @@ void DrawNode::ensureCapacityGLPoint(int count)
         _bufferCapacityPoint += MAX(_bufferCapacityPoint, count);
         _bufferGLPoint = (V2F_C4B_T2F*)realloc(_bufferGLPoint, _bufferCapacityPoint*sizeof(V2F_C4B_T2F));
         
-        _customCommandPoint.createVertexBuffer(sizeof(V2F_C4B_T2F), _bufferCapacityPoint, CustomCommand::BufferUsage::STATIC);
-        _customCommandPoint.updateVertexBuffer(_bufferGLPoint, _bufferCapacityPoint*sizeof(V2F_C4B_T2F));
+        _customCommandPoint.createVertexBuffer(sizeof(V2F_C4B_T2F), _bufferCapacityPoint, CustomCommand::BufferUsage::STATIC, _bufferGLPoint);
```
Comment (maximemontasheri) [inline cocos2d/cocos/2d/CCDrawNode.cpp]: maybe make this buffer DYNAMIC

File: cocos2d/cocos/2d/CCDrawNode.cpp [hunk 2]
```diff
@@ -126,8 +124,7 @@ void DrawNode::ensureCapacityGLLine(int count)
         _bufferCapacityLine += MAX(_bufferCapacityLine, count);
         _bufferGLLine = (V2F_C4B_T2F*)realloc(_bufferGLLine, _bufferCapacityLine*sizeof(V2F_C4B_T2F));
         
-        _customCommandLine.createVertexBuffer(sizeof(V2F_C4B_T2F), _bufferCapacityLine, CustomCommand::BufferUsage::STATIC);
-        _customCommandLine.updateVertexBuffer(_bufferGLLine, _bufferCapacityLine*sizeof(V2F_C4B_T2F));
+        _customCommandLine.createVertexBuffer(sizeof(V2F_C4B_T2F), _bufferCapacityLine, CustomCommand::BufferUsage::STATIC, _bufferGLLine);
```
Comment (maximemontasheri) [inline cocos2d/cocos/2d/CCDrawNode.cpp]: maybe DYNAMIC here too ?

File: cocos2d/cocos/2d/CCSprite.cpp [hunk 3]
```diff
@@ -462,7 +462,7 @@ void Sprite::setMVPMatrixUniform()
         if (programState) {
             updateProgramState();
             if (_mvpMatrixLocation) {
-                assert(_mvpMatrixLocation.location[0] < 20 && _mvpMatrixLocation.location[1] < 20);
+                //assert(_mvpMatrixLocation.location[0] < 20 && _mvpMatrixLocation.location[1] < 20);
```
Comment (maximemontasheri) [inline cocos2d/cocos/2d/CCSprite.cpp]: let's review this together live

File: cocos2d/cocos/renderer/backend/opengl/BufferGL.cpp [hunk 4]
```diff
@@ -127,6 +127,7 @@ void BufferGL::updateData(void* data, unsigned long size)
         }
         CHECK_GL_ERROR_DEBUG();
         _bufferAllocated = static_cast<unsigned int>(size);
+        _allocated = true;
```
Comment (maximemontasheri) [inline cocos2d/cocos/renderer/backend/opengl/BufferGL.cpp:130]: this needs to be added to metal backend

File: cocos2d/cocos/2d/CCDrawNode.cpp (hunk 1 above)
Comment (lagoPS) [reply cocos2d/cocos/2d/CCDrawNode.cpp]: Fixed

File: cocos2d/cocos/2d/CCDrawNode.cpp (hunk 2 above)
Comment (lagoPS) [reply cocos2d/cocos/2d/CCDrawNode.cpp]: Fixed!

File: cocos2d/cocos/2d/CCSprite.cpp (hunk 3 above)
Comment (lagoPS) [reply cocos2d/cocos/2d/CCSprite.cpp]: Removed check altogether

File: cocos2d/cocos/renderer/backend/opengl/BufferGL.cpp (hunk 4 above)
Comment (lagoPS) [reply cocos2d/cocos/renderer/backend/opengl/BufferGL.cpp:130]: Fixed in BufferMT

**Pull request merged**